engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/desugar/LambdaShorthandToLambdaMini.scala

```diff
@@ -0,0 +1,380 @@
+package org.enso.compiler.pass.desugar
+
+import org.enso.compiler.context.{CompilerContext, FreshNameSupply}
+import org.enso.compiler.core.{CompilerError, IR}
+import org.enso.compiler.core.ir.expression.{Application, Case, Operator}
+import org.enso.compiler.core.ir.{
+  CallArgument,
+  DefinitionArgument,
+  Expression,
+  Function,
+  IdentifiedLocation,
+  Name,
+  ProcessingPass
+}
+import org.enso.compiler.pass.{IRPass, MiniIRPass}
+
+class LambdaShorthandToLambdaMini(
+  protected val freshNameSupply: FreshNameSupply
+) extends MiniIRPass[ShouldSkipMeta] {
+
+  override def prepare(ir: IR): Unit = {
+    ir match {
+      case Application.Prefix(fn, args, _, _, _) =>
+        args.foreach {
+          case CallArgument.Specified(_, blank: Name.Blank, _, _) =>
+            blank.passData.update(this, new ShouldSkipMeta(true))
+          case _ => ()
+        }
+        fn match {
+          case blank: Name.Blank =>
+            blank.passData.update(this, new ShouldSkipMeta(true))
+          case _ => ()
+        }
+      case Application.Sequence(items, _, _) =>
+        items.foreach {
+          case blank: Name.Blank =>
+            blank.passData.update(this, new ShouldSkipMeta(true))
+          case _ => ()
+        }
+      case Case.Expr(scrutinee, _, _, _, _) =>
+        scrutinee match {
+          case blank: Name.Blank =>
+            blank.passData.update(this, new ShouldSkipMeta(true))
+        }
+      case _ => ()
+    }
+  }
+
+  override def transformIr(ir: IR): IR = {
+    ir match {
+      case name: Name          => desugarName(name)
+      case app: Application    => desugarApplication(app)
+      case caseExpr: Case.Expr => desugarCaseExpr(caseExpr)
+      case _                   => ir
+    }
+  }
+
+  private def getMeta(ir: IR): Option[ShouldSkipMeta] = {
+    ir.passData.get(this) match {
+      case Some(meta: ShouldSkipMeta) => Some(meta)
+      case _                          => None
+    }
+  }
+
+  /** Desugars an arbitrary name occurrence, turning isolated occurrences of
+    * `_` into the `id` function.
+    *
+    * @param name the name to desugar
+    * @return `name`, desugared where necessary
+    */
+  private def desugarName(name: Name): Expression = {
+    name match {
+      case blank: Name.Blank =>
+        getMeta(blank) match {
+          case Some(meta) if meta.shouldSkip =>
+            // Unset the metadata and skip
+            blank.passData.remove(this)
```

Comment:
`ShouldSkipMeta` is less than ideal _"functional pattern"_ to start with:
- `IR` is immutable (good) except `passData` (acceptable, but...)
- we mutate it on the way down in `prepare`
- we remove the modification later

E.g. instead of using `IR` as immutable, or _additive_ (new data are being added to `passData`) we use it to store local state and communicate between `prepare` and `transform`.

I believe this clearly shows why [prepare is unsufficient and better API is needed](https://github.com/enso-org/enso/pull/11191/files#r1778478784).

Reply:
Thank you.